htmresearch/frameworks/nlp/classify_document_fingerprint.py

```diff
@@ -0,0 +1,306 @@
+    """
+    Issue a reset signal to the model. The assumption is that a sequence has
+    just ended and a new sequence is about to begin.  The default behavior is
+    to do nothing - not all subclasses may re-implement this.
+    """
+    # TODO: Introduce a consistent reset method name
+    for r in self.learningRegions:
+      if r.type == 'py.TemporalPoolerRegion':
+        r.executeCommand(['reset'])
+      elif r.type == 'py.TPRegion':
+        r.executeCommand(['resetSequenceStates'])
+
+
+  def trainText(self, token, labels, sequenceId=None, reset=0):
+    """
+    Train the model with the given text token, associated labels, and
+    sequence ID.
+
+    @param token      (str)  The text token to train on
+    @param labels     (list) A list of one or more integer labels associated
+                             with this token. If the list is empty, the
+                             classifier will not be trained.
+    @param sequenceId (int)  An integer ID associated with this token and its
+                             sequence (document).
+    @param reset      (int)  Should be 0 or 1. If 1, assumes we are at the
+                             beginning of a new sequence.
+    """
+
+    # Accumulate text
+    if self.currentDocument is None:
+      self.currentDocument = [token]
+    else:
+      self.currentDocument.append(token)
+
+    # If reset issued, train on this document
+    if reset == 1:
+      document = " ".join(self.currentDocument)
+      sensor = self.sensorRegion.getSelf()
+      sensor.addDataToQueue(document, labels, sequenceId, reset)
+
+      for region in self.learningRegions:
+        region.setParameter("learningMode", True)
+      self.network.run(1)
+
+      self.currentDocument = None
+
+      # Print the outputs of each region
+      if self.verbosity >= 2:
+        print "Training with document:",document
+        self.printRegionOutputs()
+
+
+  def classifyText(self, token, reset=0):
+    """
+    Classify the token and return a list of the best classifications.
+
+    @param token    (str)  The text token to train on
+    @param reset    (int)  Should be 0 or 1. If 1, assumes we are at the
+                           end of a sequence. A reset signal will be issued
+                           after the model has been trained on this token.
+
+    @return  (numpy array) An array of size numLabels. Position i contains
+                           the likelihood that this sample belongs to the
+                           i'th category. An array containing all zeros
+                           implies no decision could be made.
+    """
+    # Accumulate text
+    if self.currentDocument is None:
+      self.currentDocument = [token]
+    else:
+      self.currentDocument.append(token)
+
+    # If reset issued, classify this document
+    if reset == 1:
+      for region in self.learningRegions:
+        region.setParameter("learningMode", False)
+        region.setParameter("inferenceMode", True)
```

Comment:
`self.learningRegions` is populated w/ regions that have a `learningMode`. This assumes those regions also have an `inferenceMode`, and will fail if they don't.